p2p/switch.go

```diff
@@ -370,6 +370,13 @@ func (sw *Switch) stopAndRemovePeer(peer Peer, reason interface{}) {
 	// https://github.com/tendermint/tendermint/issues/3338
 	if sw.peers.Remove(peer) {
 		sw.metrics.Peers.Add(float64(-1))
+	} else {
+		// Removing the peer has failed so we set a flag to mark that a removal was attempted
+		// This is done to avoid the corner case of adding a peer when this error happens before the
+		// AddPeer routine has finished. Setting this flag will prevent a peer from being added
+		// to a node's peer set.
+		peer.SetRemovalFailed()
```

Comment:
Is it safe to do this here without the peer set being locked in the same way it is when calling `GetRemovalFailed` [here](https://github.com/tendermint/tendermint/pull/9500/files#diff-9017d44e6c3e27f96079fbe86494298ad63faef9c9c615fe2f499dd5a3765c34R50)? (the `GetRemovalFailed` call happens while the peer set's mutex is locked)

Reply:
I had intuitively implemented with a lock in a previous version (although by setting a lock on the peer itself, not on the entire set). But I did not notice the attack being more successful so I opted to not add a mutex into the peer structure itself even though I still do a small window of opportunity for the removal to happen first and then the Add to grab a lock and add the peer before we set the removal failed flag. Thinking about it now, I could actually set the flag inside the Remove function above the line that returns [false](https://github.com/tendermint/tendermint/blob/499804c6232f372a2ff1d753889f7d5c8f8b33c3/p2p/peer_set.go#L113) and not do a lock on the peer itself. I will do a local test with this change to examine the behaviour unless you have objections.

Reply:
Based on a local test I think it is better to put it inside the Remove. Will confirm on the testnet before pushing the change.